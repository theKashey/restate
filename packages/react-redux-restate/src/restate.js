import React, {Component} from 'react';
import PropTypes from 'prop-types';
import shallowequal from 'shallowequal';
import {createProvider} from 'react-redux';
import reduxRestate from 'redux-restate';
import hoistStatics from 'hoist-react-statics';
import displayName from 'react-display-name';

const getStores = (stores, props, context) => {
  const result = {};
  Object.keys(stores).forEach(key => {
    const value = stores[key];
    if (typeof value === 'string') {
      const store = props[value] || context[value];
      if (store) {
        result[key] = store;
      } else {
        throw new Error(`restate: unable to found store "${value}" for key "${key}"`);
      }
    } else {
      result[key] = value;
    }
  });
  return result;
};

const nullFn = () => ({});

const noChildren = {children: undefined};

const restate = (baseStores, composeState, routeDispatch, options = nullFn) => WrappedComponent => {
  if(options && typeof options !== 'function') {
    throw new Error('react-redux-restate: options should be an option')
  }
  const basicOptions = options({});
  const storeKey = basicOptions.storeKey || 'store';
  const restateKey = basicOptions.restateKey || 'store';

  const Provider = createProvider(restateKey);

  const contextTypes = {
    [storeKey]: PropTypes.any,
  };

  Object.keys(baseStores).forEach(key => {
    const value = baseStores[key];
    if (typeof value === 'string') {
      contextTypes[value] = PropTypes.any;
    }
  });

  const ignoredProps = {};
  const deeperProps = basicOptions.deeperProps || [];

  (basicOptions.ignoreProps || []).forEach(prop => {ignoredProps[prop] = undefined});

  const compareProps = (nextProps, currentProps, additionalIgnore = {}) => (
    shallowequal({...nextProps, ...ignoredProps, ...additionalIgnore}, {...currentProps, ...ignoredProps, ...additionalIgnore})
    && deeperProps.reduce((acc, line) => acc && shallowequal(nextProps[line], currentProps[line]), true)
  );

  class RestateComponent extends Component {
    static contextTypes = contextTypes;

    constructor(props, context) {
      super();
      this.stores = {
        ...getStores(baseStores, props, context),
        default: props[storeKey] || context[storeKey],
      };
      this.propsOverride = null;
    }

    componentWillMount() {
      this.store = reduxRestate(this.stores, this.composeState, this.routeDispatch, this.getOptions());
    }

    componentWillReceiveProps(nextProps) {
      if (!compareProps(nextProps, this.props, noChildren)) {
        this.propsOverride = nextProps;
        this.store.replaceOptions(options(nextProps));
        this.store.update();
        this.propsOverride = null;
      }
    }

    shouldComponentUpdate(nextProps) {
      return !compareProps(nextProps, this.props);
    }

    componentWillUnmount() {
      this.store.unsubscribe();
    }

    getOptions() {
      return options(this.props);
    }

    composeState = stores => composeState(stores, this.propsOverride || this.props);
    routeDispatch = (dispatches, event) => routeDispatch(dispatches, event, this.propsOverride || this.props);

    render() {
      return (
        <Provider store={this.store}>
          <WrappedComponent {...this.props} />
        </Provider>
      );
    }
  }

  hoistStatics(RestateComponent, WrappedComponent);
  RestateComponent.displayName = `reconnected(${displayName(WrappedComponent)})`;

  return RestateComponent;
};

export default restate;
