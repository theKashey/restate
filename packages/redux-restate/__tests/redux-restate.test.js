import storeMock from 'redux-mock-store';
import restate from '../src';

describe('redux-restate', () => {
  const createStore = storeMock([]);

  it('smoke test', () => {
    let storeData1 = {
      branch1: { a: 1 },
      brancha: {},
    };

    const storeData2 = {
      branch1: { b: 2 },
      branchb: {},
    };

    const store1 = createStore(() => storeData1);
    const store2 = createStore(() => storeData2);

    const restore = restate(
      {
        store1,
        store2,
      },
      ({ store1: state1, store2: state2 }) => ({
        ...state1,
        ...state2,
        soil: 42,
      }),
      ({ store1: dispatch1, store2: dispatch2 }, event) => {
        if (event.store === 1) {
          dispatch1(event);
        } else {
          dispatch2(event);
        }
      },
    );

    expect(restore.getState()).toEqual({
      branch1: { b: 2 },
      brancha: {},
      branchb: {},
      soil: 42,
    });

    const listener1 = jest.fn();
    const listener2 = jest.fn();
    const listenerr = jest.fn();

    store1.subscribe(listener1);
    store2.subscribe(listener2);

    const unsubscribe1 = restore.subscribe(listenerr);

    store1.dispatch({ type: 'event', id: 1 });
    expect(listener1).toHaveBeenCalledTimes(1);
    expect(listener2).toHaveBeenCalledTimes(0);
    expect(listenerr).toHaveBeenCalledTimes(0);

    storeData1 = {
      ...storeData1,
      change: 1,
    };

    store1.dispatch({ type: 'event', id: 1 });
    expect(listenerr).toHaveBeenCalledTimes(1);
    store2.dispatch({ type: 'event', id: 1 });
    expect(listenerr).toHaveBeenCalledTimes(1);

    unsubscribe1();

    storeData1 = {
      ...storeData1,
      change: 2,
    };

    store2.dispatch({ type: 'event', id: 1 });
    expect(listener1).toHaveBeenCalledTimes(2);
    expect(listener2).toHaveBeenCalledTimes(2);
    expect(listenerr).toHaveBeenCalledTimes(1);

    restore.subscribe(listenerr);

    restore.dispatch({ type: 'event', store: 1 });
    expect(listener1).toHaveBeenCalledTimes(3);
    expect(listener2).toHaveBeenCalledTimes(2);

    restore.dispatch({ type: 'event', store: 2 });
    expect(listener1).toHaveBeenCalledTimes(3);
    expect(listener2).toHaveBeenCalledTimes(3);

    expect(store1.getActions()).toMatchSnapshot();
    expect(store2.getActions()).toMatchSnapshot();
  });

  it('should stop focued updates', () => {});
});
