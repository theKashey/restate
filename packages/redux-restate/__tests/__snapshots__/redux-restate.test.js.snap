// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux-restate smoke test 1`] = `
Array [
  Object {
    "id": 1,
    "type": "event",
  },
  Object {
    "store": 1,
    "type": "event",
  },
]
`;

exports[`redux-restate smoke test 2`] = `
Array [
  Object {
    "id": 1,
    "type": "event",
  },
  Object {
    "store": 2,
    "type": "event",
  },
]
`;
